Реализовываем анимацию заголовков на экране «Результат игры»

a1re opened the pull request:
**Добавлена анимация заголовков победы и проигрыша**

Саму анимацию можно пока увидеть только в svg-файлах, т.к. я не понял, можно ли вообще сейчас вызвать победу или поражение на самом сайте. Если все-таки можно, то следует иметь в виду, что высота холста заголовка поражения увеличена в два раза, чтобы обеспечить анимацию падения букв.

---
:mortar_board: [Реализовываем анимацию заголовков на экране «Результат игры»](https://up.htmlacademy.ru/animation/1/user/1339161/tasks/12)

:boom: https://htmlacademy-animation.github.io/1339161-magic-vacation-1/10/

a1re committed:
feat: Добавлена анимация заголвов победы и проигрыша

Саму анимацию можно пока увидеть только в svg-файлах, т.к. я не
понял, можно ли вообще сейчас вызвать победу или поражение на
самом сайте. Если все-таки можно, то следует иметь в виду, что
высота холста заголовка поражения увеличена в два раза, чтобы
обеспечить анимацию падения букв.

sfi0zy commented on source/img/lose-heading.svg:
```diff
@@ -1,25 +1,193 @@
+        fill="freeze"/>
+      <animate
+        attributeName="stroke-dasharray"
+        from="1 100" to="100 0"
```
:red_circle: - Длина контура должна вычисляться точно, через метод getTotalLength(). Критерий Б23.

:point_up: - Заголовки уже встроены в страницу, а не используются в виде отдельных файлов. Это не случайность. Когда SVG встроена - мы можем оперировать всеми элементами в ней с помощью скриптов. Это удобно. И с этим стоит попрактиковаться.

:point_up: - Также щепотка JS решит вопрос с копипастой практически одинаковых анимаций для каждой отдельной буквы.

a1re replied:
> Длина контура должна вычисляться точно, через метод getTotalLength()

Я так сначала и сделал, но обратил внимание, что в референсах анимация немного другая — она рисуется из нескольких точек, а не из одной. Это логичнее анимировать уменьшением пропуска в stroke-dasharray, а не stroke-dashoffset, как в демонстрации из учебника. Как поступить?

sfi0zy replied:
> > Длина контура должна вычисляться точно, через метод getTotalLength()

> Я так сначала и сделал, но обратил внимание, что в референсах анимация немного другая — она рисуется из нескольких точек, а не из одной. Это логичнее анимировать уменьшением пропуска в stroke-dasharray, а не stroke-dashoffset, как в демонстрации из учебника. Как поступить?

Так одно другому не мешает - подумайте, откуда взялись значения в stroke-dasharray? С чем они связаны и почему вы взяли именно значение 100?

a1re replied:
Понял, спасибо)

sfi0zy reviewed (commented):
:pencil2: **Первичная проверка.**

**Обозначения:**

:red_circle: - Нужно обязательно исправить.
:point_up: - Нужно подумать над этим, принять к сведению.
:watch: - Нужно подождать (что-то пошло не так, вы ни при чем).
:heavy_check_mark: - Задание принято.
:x: - Задание не принято.
:large_blue_diamond: - Важное сообщение.

**Пункты задания:**

- [ ] 1. Необходимо сделать анимацию рисования контуров заголовков результатов игры «Победа!» и «Не угадал». Анимация начинает проигрываться во время отображения Результата игры после истечение таймера. Обратите внимание, что рисуется из двух-трех точек одновременно.
- [x] 2. Символы «Не угадал!» при этом должны падать сверху с одним отскоком.
- [x] 3. Все символы «Победа!» появляются одновременно, а символы «Не угадал!» должны появляться с задержкой внахлёст. Обратите внимание на движение контура, которых создают символы вместе: он изогнутый, не прямой.

**Комментарии:**

> Саму анимацию можно пока увидеть только в svg-файлах, т.к. я не понял, можно ли вообще сейчас вызвать победу или поражение на самом сайте.

Да, можно, загляните в result.js. Там есть временное решение для этого (уберите display: none у кнопок, чтобы их увидеть и кликать по ним).

:red_circle: - Анимация должна отображаться на экранах с результатами игры.

:large_blue_diamond: - Это последнее задание, связанное с SVG и SMIL-анимациями. Дальше мы перейдем к другим темам.

**Итог:**

:x: - Выполнены не все требования ТЗ, не соблюдены базовые критерии. Необходима доработка.

a1re committed:
feat: Добавлена интеграция анимации заголовков в верстку

Интеграция SMIL-анимаций полностью перенесено из исходного кода
SVG в JS (файл result-animation.js). Это сделано для того, чтобы
можно было запускать анимацию вручную в момент появления
заголовка — при этом у положительного результата фактически два
одинаковых заголовка, а id анимаций должно быть разным. Длина
контура расчитывается динамически для каждой буквы.

a1re commented:
@sfi0zy интегрировал заголовки в верстку, сделал расчет длины контура как требуется по критерию и динамическое добавление анимации «падения» букв.

Пришлось прямо сильно всё переделать, но надеюсь теперь правильно.

Хорошая тренировка.

sfi0zy commented on source/js/additions/result-animation.js:
```diff
@@ -0,0 +1,164 @@
+    const heading = document.querySelector(selector + ` .positiveHeading`);
+
+    const scale = document.createElementNS(`http://www.w3.org/2000/svg`, `animateTransform`);
+    scale.setAttribute(`id`, id + ANIMATION_ID_POSTFIX);
+    scale.setAttribute(`attributeName`, `transform`);
+    scale.setAttribute(`type`, `scale`);
+    scale.setAttribute(`from`, APPEAR_SCALE_FROM);
+    scale.setAttribute(`to`, APPEAR_SCALE_TO);
+    scale.setAttribute(`begin`, `indefinite`);
+    scale.setAttribute(`dur`, APPEAR_SCALE_DURATION);
+    scale.setAttribute(`fill`, `freeze`);
```
:point_up: - Вот этот кусок кода повторяется несколько раз, может для этой задачи какую-то отдельную функцию сделать?

sfi0zy reviewed (commented):
:pencil2: **Повторная проверка.**

**Пункты задания:**

- [x] 1. Необходимо сделать анимацию рисования контуров заголовков результатов игры «Победа!» и «Не угадал». Анимация начинает проигрываться во время отображения Результата игры после истечение таймера. Обратите внимание, что рисуется из двух-трех точек одновременно.
- [x] 2. Символы «Не угадал!» при этом должны падать сверху с одним отскоком.
- [x] 3. Все символы «Победа!» появляются одновременно, а символы «Не угадал!» должны появляться с задержкой внахлёст. Обратите внимание на движение контура, которых создают символы вместе: он изогнутый, не прямой.

**Комментарии:**

> Пришлось прямо сильно всё переделать, но надеюсь теперь правильно. Хорошая тренировка.

:muscle: - Ну будем считать, что с SVG вы разобрались, попробовали взаимодействовать с картинками в разных контекстах. Это хорошо. Дальше вас ждет канвас, и, забегая вперед, могу сказать, что каждый следующий модуль будет больше предыдущего, так что приготовьтесь, самая жесть еще впереди.

**Итог:**

:heavy_check_mark: :+1: - Выполнены все требования ТЗ, базовые критерии соблюдены, анимация соответствует видео с концептом проекта. Задание принято.
